### PR TITLE
feat: open smartlock (only nuki is supported)

### PR DESCRIPTION
**Pull request by kuamanet**

### CODE REVIEW

File: web-api/src/Kerbero.Domain/SmartLocks/Dtos/OpenSmartLockParams.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+public class OpenSmartLockParams
+{
+  public required Guid UserId { get; set; }
+
+  public required SmartlockProvider Provider { get; set; }
+
+  public required string SmartLockId { get; set; }
+
+  public int CredentialsId { get; set; }
```
Comment (vash72) [inline web-api/src/Kerbero.Domain/SmartLocks/Dtos/OpenSmartLockParams.cs:14]:
```suggestion
  public int CredentialId { get; set; }
```

File: web-api/src/Kerbero.Domain/SmartLocks/Dtos/OpenSmartLockParams.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+using System;
+using Kerbero.Domain.SmartLocks.Models;
+
+namespace Kerbero.Domain.SmartLocks.Dtos;
+
+public class OpenSmartLockParams
+{
+  public required Guid UserId { get; set; }
+
+  public required SmartlockProvider Provider { get; set; }
```
Comment (vash72) [inline web-api/src/Kerbero.Domain/SmartLocks/Dtos/OpenSmartLockParams.cs:10]:
```suggestion
  public required SmartlockProvider SmartlockProvider { get; set; }
```

File: web-api/src/Kerbero.Domain/SmartLocks/Interactors/OpenSmartLockInteractor.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        UserId = openSmartLockParams.UserId,
+      });
+
+    if (userNukiCredentialsResult.IsFailed)
+    {
+      return Result.Fail(userNukiCredentialsResult.Errors);
+    }
+
+    // ensure provided credentials belong to current user
+    var credentials = userNukiCredentialsResult.Value.Find(it => it.Id == openSmartLockParams.CredentialsId);
```
Comment (vash72) [inline web-api/src/Kerbero.Domain/SmartLocks/Interactors/OpenSmartLockInteractor.cs:44]:
```suggestion
    var credentialBelongToUser  = userNukiCredentialsResult.Value.Any(it => it.Id == openSmartLockParams.CredentialsId);
```

File: web-api/src/Kerbero.Domain/SmartLocks/Interactors/OpenSmartLockInteractor.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (credentials is null)
+    {
+      return Result.Fail(new UnauthorizedAccessError());
+    }
+
+    if (openSmartLockParams.Provider != SmartlockProvider.Nuki)
+    {
+      return Result.Fail(new UnsupportedProviderError());
```
Comment (vash72) [inline web-api/src/Kerbero.Domain/SmartLocks/Interactors/OpenSmartLockInteractor.cs:53]:
```suggestion
      return Result.Fail(new UnsupportedSmartLockProviderError());
```
Provider seems to be too generic

File: web-api/src/Kerbero.WebApi/Controllers/SmartLocksController.cs [hunk 5]
```diff
@@ -55,4 +61,36 @@ public async Task<ActionResult<List<SmartLockDto>>> GetAll()
...
+    var provider = SmartlockProvider.TryParse(request.Provider);
+
+    if (provider is null)
+    {
+      return BadRequest();
+    }
+
+    var userId = HttpContext.GetAuthenticatedUserId();
+
+    // try to open the smartlock
```
Comment (vash72) [inline web-api/src/Kerbero.WebApi/Controllers/SmartLocksController.cs:79]:
```suggestion

```
the code below is clear enough

**Review by vash72 — CHANGES_REQUESTED** [review]
i like the current state of SmartLockProvider

File: web-api/src/Kerbero.Domain/SmartLocks/Models/SmartlockProvider.cs [hunk 6]
```diff
@@ -6,11 +6,21 @@ namespace Kerbero.Domain.SmartLocks.Models;
 public class SmartlockProvider
```
Comment (salvanicola) [inline web-api/src/Kerbero.Domain/SmartLocks/Models/SmartlockProvider.cs:6]: Lo metterei in common rinominato in Provider

File: web-api/src/Kerbero.Domain/SmartLocks/Interactors/OpenSmartLockInteractor.cs (hunk 3 above)
Comment (kuamanet) [reply web-api/src/Kerbero.Domain/SmartLocks/Interactors/OpenSmartLockInteractor.cs:44]: I need the credentials to pass them over the repository

**Review by vash72 — DISMISSED** [review]

File: web-api/src/Kerbero.WebApi/Models/Requests/OpenSmartLockRequest.cs [hunk 7]
```diff
@@ -0,0 +1,3 @@
+namespace Kerbero.WebApi.Models.Requests;
+
+public record OpenSmartLockRequest(int CredentialsId, string SmartLockProvider);
```
Comment (vash72) [inline web-api/src/Kerbero.WebApi/Models/Requests/OpenSmartLockRequest.cs:3]:
```suggestion
public record OpenSmartLockDto(int CredentialsId, string SmartLockProvider);
```

**Review by vash72 — CHANGES_REQUESTED** [review]

**Review by vash72 — APPROVED** [review]

**Pull request merged**